### ISSUE TITLE
Using trial -j can result in wrong paths in sys.path

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6390]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#6390|
|Type|defect|
|Created|2013-03-26 11:15:14Z|
|Branch|https://github.com/twisted/twisted/tree/dist-trial-path-6390|

The _preamble module in twisted/trial/_dist is fine for non-system install of Twisted, but it can mess up sys.path the way it calculates the Twisted path on system installs. We should remove that.

<details><summary>Searchable metadata</summary>

```
trac-id__6390 6390
type__defect defect
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_dist_trial_path_6390 branches/dist-trial-path-6390
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1364296514000000 1364296514000000
changetime__1369878732000000 1369878732000000
version__None None
owner__therve therve
cc__jml cc__free.ekanayaka
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [37809]) Branching to 'dist-trial-path-6390'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

I tried something in the branch, by passing the direct value of sys.path as a separate environment variable to the workers. Using some tests it seemed that PYTHONPATH could result in weird order issues.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Can you describe how this breaks things a bit more?  Right now it's not clear to me what bug I should expect to be fixed in the branch.  Although I do like the idea of removing this copy of _preamble.py for its own sake.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

Sure. As a concrete example, using the twisted package on Ubuntu means that /usr/share/pyshared ends up in sys.path because that's where the .py files are, so it can confuse things up. trial itself is protected from that because we don't ship bin/_preamble, so the import fails and things go as normal, but we do ship the one in twisted/trial/_dist.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_7"></a>|@tomprince set owner to @therve|
|-|-|

Is there any way to test this without needing to touch `/usr`? I'd like to set something up on buildbot to test this, but there aren't enough details here to figure out how to do that.

Presumably, there is other global state (I'm thinking of `python -Wall` is also not preserved.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

There is certainly a way, maybe mangling sys.path before calling trial. I don't think it's worth having a buildbot for that though (even considering we don't even have a buildbot using trial -j).

Maybe you can open a bug about other python flags not being respected.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/dist-trial-path-6390).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to @therve|
|-|-|

Thanks, therve.

  1. I think [these pydoctor errors](http://buildbot.twistedmatrix.com/builders/documentation/builds/3318/steps/api-documentation/logs/new%20pydoctor%20errors) are spurious, but please make sure.
  2. I am unnerved by the fact that `TRIAL_PYTHONPATH` is now a wholly undocumented implementation detail; a trick made possible only by the fact that adding it doesn't involve adding any top-level suites.  Please add some documentation to the module docstring of `workertrial` and the docstring of `launchWorkerProcess` explaining this side-effect.  (Maybe a `@see` referencing each other so it's easy to find where the variable comes from / goes to.)
  3. The `arguments` variable in the test case is already really misleading (why not `arguments.extend(args)`?) and this change just makes it truly awful :).  Please relay the value of `env` to the outer scope in a more readable way.
  4. There's no test coverage for the positive branch of `workertrial`, so nothing can verify the integration of the path setup.  If you could find a way to put at least part of this into a function and test it that would be great.

These are all relatively minor points, so please address to your satisfaction and land.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [38491]) Merge dist-trial-path-6390

Author: therve
Reviewers: glyph, tom.prince
Fixes: [#6390](https://github.com/twisted/twisted/issues/6390)

Remove the preamble setting the sys.path for trial dist children, and pass an
environment variable instead.
